Sources/SwiftParser/Directives.swift

```diff
@@ -86,14 +86,20 @@ extension Parser {
       while let poundIfHandle = self.canRecoverTo(any: firstIteration ? [.poundIfKeyword] : [.poundIfKeyword, .poundElseifKeyword, .poundElseKeyword]),
         loopProgress.evaluate(self.currentToken)
       {
-        let (unexpectedBeforePoundIf, poundIf) = self.eat(poundIfHandle)
+        var (unexpectedBeforePoundIf, poundIf) = self.eat(poundIfHandle)
+        var unexpectedBetweenPoundKeywordAndCondition: RawUnexpectedNodesSyntax?
         firstIteration = false
         // Parse the condition.
         let condition: RawExprSyntax?
         switch poundIf.tokenKind {
         case .poundIfKeyword, .poundElseifKeyword:
           condition = RawExprSyntax(self.parseSequenceExpression(.basic, forDirective: true))
         case .poundElseKeyword:
+          if self.currentToken.rawTokenKind == .keyword(.if),
+              let poundElseHandle = self.canRecoverTo(.poundElseKeyword) {
+            (unexpectedBetweenPoundKeywordAndCondition, _) = self.eat(poundElseHandle)
+            poundIf = self.missingToken(.poundElseifKeyword)
+          }
```

Comment:
I see three issues with this approach

- We eat all tokens up to the next the next `#else` as unexpected tokens when you call `self.eat(poundElseHandle)`. We should be treating the `#else if` as if it was spelled as `#elseif` parse anything in between the `#else if` and `#else` as normal elements in the `IfConfigClauseSyntax`
- AFAICT we are dropping the `#else` you parse with `self.eat(poundElseHandle)` on the floor and thus loosing [source fidelity](https://swiftpackageindex.com/apple/swift-syntax/main/documentation/swiftparser/parserdesign#Source-Fidelity). I’m surprised this isn’t caught by the test case you added.

I think what you want to do instead is:
- Check whether the currentToken is an `if` keyword (you can use `self.at(.keyword(.if))` for that, which is slightly shorter than what you wrote but functionally equivalent)
- If you discover that case, you should shove the `#else` and `if` into `unexpectedBeforePoundIf` so they are not lost in the syntax tree and we maintain source fidelity
- Synthesize the `poundIf` as a `poundElseifKeyword` like you already do.

Reply:
Regarding your suggestions:
1. > Check whether the currentToken is an `if`

I'm using the `consume(if:)` method, which returns a `RawTokenSyntax` I can inform as unexpected (item 2):
```swift
if let ifToken = self.consume(if: .keyword(.if), allowTokenAtStartOfLine: false) {
```

2. > you should shove the `#else` and `if` into `unexpectedBeforePoundIf`

The only way I figured out how to do this was by using `RawUnexpectedNodesSyntax` combining initializers:
```swift
unexpectedBeforePoundIf = RawUnexpectedNodesSyntax(combining: unexpectedBeforePoundIf, poundIf, ifToken, arena: self.arena)
```

Is this acceptable?

Reply:
Yes, both solutions not only acceptable but the right way to go. 👍